be/src/storage/dictionary_cache_manager.h

```diff
@@ -57,78 +73,186 @@ struct DictionaryCacheHashTraits {
 
     using KeyCppType = typename DictionaryCacheTypeTraits<KeyLogicalType>::CppType;
     using ValueCppType = typename DictionaryCacheTypeTraits<ValueLogicalType>::CppType;
 
-    virtual inline Status insert(const Datum& k, const Datum& v) override {
-        KeyCppType key;
-        ValueCppType value;
-        _get<KeyCppType>(k, key);
-        _get<ValueCppType>(v, value);
-        auto r = _dictionary.insert({key, value});
-        if (!r.second) {
-            return Status::InternalError("duplicate key found when refreshing dictionary");
-        }
-        _total_memory_useage.fetch_add(_get_element_memory_usage<KeyCppType, ValueCppType>(key, value));
-        return Status::OK();
-    }
+    virtual inline Status insert(const Datum& k, const Datum& v, const uint8_t& flag) override {
+        switch (_type) {
+        case DictionaryCacheEncoderType::PK_ENCODE: {
+            KeyCppType key;
+            ValueCppType value;
+
+            // Memory structure for string type value:
+            //  Slice 1 -> |  buffer  |
+            //             |buffer size|
+            //  Slice 2 -> |  buffer  |
+            //             |buffer size|
+            //  Slice 3 -> |  buffer  |
+            //             |buffer size|
+            if constexpr (std::is_same_v<KeyCppType, Slice>) {
+                const Slice& input = k.get<KeyCppType>();
+                auto* buffer = reinterpret_cast<char*>(_pool.allocate(input.size));
+                RETURN_IF_UNLIKELY_NULL(buffer, Status::MemoryAllocFailed("alloc mem for dictionary failed"));
+                memcpy(buffer, input.data, input.size);
+                key.data = buffer;
+                key.size = input.size;
+            } else {
+                key = k.get<KeyCppType>();
+            }
 
-    virtual inline Status lookup(const Datum& k, Column* dest) override {
-        KeyCppType key;
-        _get<KeyCppType>(k, key);
-        auto iter = _dictionary.find(key);
-        if (iter == _dictionary.end()) {
-            return Status::NotFound("key not found in dictionary cache");
+            // Memory structure for string type value:
+            //             | fast decode flag |
+            //  Slice 1 -> |  buffer  |
+            //             |buffer size|
+            //             | fast decode flag |
+            //  Slice 2 -> |  buffer  |
+            //             |buffer size|
+            //             | fast decode flag |
+            //  Slice 3 -> |  buffer  |
+            //             |buffer size|
+            if constexpr (std::is_same_v<ValueCppType, Slice>) {
+                const Slice& input = v.get<ValueCppType>();
+                size_t allocate_size = input.size + sizeof(uint8_t);
+                auto* buffer = reinterpret_cast<char*>(_pool.allocate(allocate_size));
+                RETURN_IF_UNLIKELY_NULL(buffer, Status::MemoryAllocFailed("alloc mem for dictionary failed"));
+                memcpy(buffer, &flag, sizeof(uint8_t));
```

Comment:
What's flag used for?

Reply:
This flag is used to indicate that if we can use fast decoding method